zephyr/lib/alloc.c

```diff
@@ -367,6 +447,7 @@ static int heap_init(void)
 
 #if CONFIG_L3_HEAP
 	sys_heap_init(&l3_heap.heap, UINT_TO_POINTER(get_l3_heap_start()), get_l3_heap_size());
+	l3_heap_flush_invalidate(&l3_heap);
```

Comment:
I'm not sure if this is safe with cached memory. sys_heap will traverse the metadata and access blocks before and after the allocated one. So even if the individual allocations are aligned to cachelines, when a block is freed/allocated, the sys_heap will access the N-1 and N+1 block metadata. These blocks might be allocated to different cores, and with a non-coherent architecture, there's no way to do this safely. I.e. if the metadata is accessed via cached memory, then there's no way to guarantee coherency --- if multiple cores access the heap. This is why we put the sys_heap metadata on uncached memory.  Right @lyakh ?

I understand this is very limiting, but can we limit L3 heap to a single core only. That would work. 

Otherwise, I don't think we can use sys_heap for L3 on these platforms, but require an alternative implementation that can handle the specific constraints we have here.

Reply:
Agree. Currently L3 heap is used only by core 0 in two cases:
- allocate D3 context storage location;
- allocate loadable modules storage;
Both operations could be done only by core 0. 
This patch is provided to unblock LNL SOF testing on RVP.  WE can consider to replace the sys_heap API with for example Memory Slabs https://docs.zephyrproject.org/latest/kernel/memory_management/slabs.html.
I could add explicit check against core number during memory allocation/free.

Reply:
no, spinlocks don't prevent incoherency, they only prevent concurrent access.

Reply:
Oh, I was thinking about combination of spinlock and correct cache flushing/invalidating. But You are right, this could be done from zephyr level when it is exactly known, which data to flush and invalidate. From SOF level it will not work with multiple core access. 
For now limiting L3_HEAP to primary core only.

Reply:
>I was thinking about combination of spinlock and correct cache flushing/invalidating

coherent.h provides such functionality